pkg/restore/restore.go

```diff
@@ -1655,22 +1656,39 @@ func (ctx *restoreContext) getOrderedResourceCollection(backupResources map[stri
 // getSelectedRestoreableItems applies Kubernetes selectors on individual items
 // of each resource type to create a list of items which will be actually
 // restored.
-func (ctx *restoreContext) getSelectedRestoreableItems(resource, targetNamespace, originalNamespace string, items []string) (restoreableResource, Result, Result) {
-	var res restoreableResource
+func (ctx *restoreContext) getSelectedRestoreableItems(resrc, targetNamespace, originalNamespace string, items []string) (restoreableResource, Result, Result) {
 	warnings, errs := Result{}, Result{}
-	res.resource = resource
-	if res.selectedItemsByNamespace == nil {
-		res.selectedItemsByNamespace = make(map[string][]restoreableItem)
+
+	restorable := restoreableResource{
+		resource: resrc,
+	}
+
+	if restorable.selectedItemsByNamespace == nil {
+		restorable.selectedItemsByNamespace = make(map[string][]restoreableItem)
 	}
 
 	if targetNamespace != "" {
-		ctx.log.Infof("Resource '%s' will be restored into namespace '%s'", resource, targetNamespace)
+		ctx.log.Infof("Resource '%s' will be restored into namespace '%s'", resrc, targetNamespace)
 	} else {
-		ctx.log.Infof("Resource '%s' will be restored at cluster scope", resource)
+		ctx.log.Infof("Resource '%s' will be restored at cluster scope", resrc)
+	}
+
+	// If the APIGroupVersionsFeatureFlag is enabled, the item path will be
+	// updated to include the API group version that was chosen for restore. For
+	// example, for "horizontalpodautoscalers.autoscaling", if v2beta1 is chosen
+	// to be restored, then "horizontalpodautoscalers.autoscaling/v2beta1" will
+	// be part of item path. Different versions would only have been stored
+	// if the APIGroupVersionsFeatureFlag was enabled during backup. The
+	// chosenGrpVersToRestore map would only be populated if
+	// APIGroupVersionsFeatureFlag was enabled for restore and the minimum
+	// required backup format version has been met.
+	cgv, ok := ctx.chosenGrpVersToRestore[resrc]
```

Comment:
I'm trying to follow what happened in your original PR and in the restore progress one that lost this section.

In #3133 when this was added, we were using [`groupResource.String()`](https://github.com/vmware-tanzu/velero/pull/3133/files#diff-e0498755be37b683528e2569667d67818bb676827c420aab49adc62cb6d47fe3R747-R759) as the key for the map lookup, but now we're using `resrc`. I think this is okay given that this function is called with `groupResource.String()` as the first parameter but I just want to make sure I'm following this correctly as it's tricky to know what these variables represent. Is my understanding correct here?

Reply:
It's good to make sure, because the code is difficult to follow. Yes, I believe your understanding is correct. Here are the lines and links that follow the logic you stated:

PR 3133 had [line 755](https://github.com/codegold79/velero/blob/00fd5ff738fc51c765b57393859bac01335ae535/pkg/restore/restore.go#L755). Exactly as you wrote, `groupResource.String()` is used as the key for the map lookup.

```go
groupResource := schema.ParseGroupResource(resource)

cgv, ok := ctx.chosenGrpVersToRestore[groupResource.String()]
if ok {
	resource = filepath.Join(groupResource.String(), cgv.Dir)
}
```

For this PR 3583, `resrc` is used instead. This variable was passed into the [`getSelectedRestoreableItems(...)` method](https://github.com/vmware-tanzu/velero/blob/ccbd30f450a0b332e3dfe5a721ba4fad65845801/pkg/restore/restore.go#L1659). This method was called in the `getOrderedResourceCollection(...)` method on [line 1643](https://github.com/vmware-tanzu/velero/blob/ccbd30f450a0b332e3dfe5a721ba4fad65845801/pkg/restore/restore.go#L1643).

```go
res, w, e := ctx.getSelectedRestoreableItems(groupResource.String(), targetNamespace, namespace, items)
```

When I ran the API group versions e2e tests, I was able to verify that the correct <resource>.<group> strings were being passed in and the path was being modified as intended.

Reply:
Thanks for the detailed explanation!